### PR TITLE
overlord: implement support for delayed restarts in the restart logic

### PR DESCRIPTION
**Pull request by Meulengracht**
This PR contains the restart logic from the primary PR (https://github.com/snapcore/snapd/pull/12718). This PR also builds on top of https://github.com/snapcore/snapd/pull/12720 to get the taskrunner hooks functionality which is needed to get informed about when a change has no more tasks to run.

The PR is in draft until I can confirm any test failures is not related to this PR and I've had a chance to go through my unit test code changes with some fresh eyes

A small overview:

### Changes to unit tests
- A lot of unit tests now need to do different checks, and some unit tests need additional explicit restart logic to simulate restarts which before was implicit. This has caused a large number of unit tests to be updated as the semantics around how restarts are now done has changed.
- If a unit test suddenly has added additional restarts, this is because the restart is no longer happening directly. This only means that the restart is now more visible in the unit test instead of having been done implicitly before.

### overlord/restart/restart.go
- The restart logic is now split into two phases. The scheduling of restart and the execution of the restarts. Before this was one single code-path, but now this happens seperately. Tasks can now request reboots, which either puts that task into WaitStatus or it puts the dependencies of that task into WaitStatus based on the context of the change.
- When a change runs out of tasks to run, the restart logic is notified, and the restart logic will check if any restarts are scheduled for that change, if it is, then it reuses the same restart logic as before and performs the reboot.

### overlord/overlord.go
- Seperated some of the logic to allow instantiation of the RestartManager after creating the taskrunner, so we can supply that as a parameter to RestartManager.Manager(), which is needed to register the hook.

### overlord/snapstate/handlers.go
- Delay/schedule the re-refresh task to wait for any pending reboot

### overlord/state/taskrunner.go
- Imported changes from PR https://github.com/snapcore/snapd/pull/12720 to support the hooks used in the restart logic

### CODE REVIEW

**Comment by Meulengracht** [comment]
I'll make sure to clean up the commits, they are a bit messy as I've switched computer twice during the development of this

**Comment by pedronis** [comment]
> overlord/snapstate/snapstate.go
Move the "system-restart-immediate" handling into the restart package

@Meulengracht do we need this change now, I'm not sure how it relates to the rest?

File: overlord/restart/restart.go [hunk 1]
```diff
@@ -299,46 +315,10 @@ func notifyRebootRequiredClassic(rebootRequiredSnap string) error {
...
-// its status and requesting a restart.
-// It should usually be invoked returning its result immediately from the
-// caller.
-// In some situations it might not set the desired status directly and schedule
-// the restart. If a manual restart is preferred (like on classic, where
-// automatic restarts are undesirable) it will instead set the task to
-// WaitStatus and return a marker error of type state.Wait.
-// The restart manager itself will then make sure to set the the status as
-// requested later on system restart to allow progress again.
-func FinishTaskWithRestart(task *state.Task, status state.Status, rt RestartType, snapName string, rebootInfo *boot.RebootInfo) error {
```
Comment (pedronis) [inline overlord/restart/restart.go:312]: why the change of name?

**Review by pedronis — COMMENTED** [review]
this is hard to follow because it seems to be doing refactoring and feature additions at the same time, and I feel a bit unsure about the refactorings, especially mixed with the rest

File: overlord/restart/restart.go (hunk 1 above)
Comment (Meulengracht) [reply overlord/restart/restart.go:312]: Originally during I wrote this, this co-existed with `FinishTaskWithRestart`, I renamed this back to reduce the number of changes.

**Comment by Meulengracht** [comment]
Closing this in preference of other PRs
